https://github.com/netty/netty/issues/3726 - libaio wrapper

clebertsuconic opened the pull request:
text from the main commit:

    [#3726] Implementation of a JNI wrapper for libaio
    
    https://github.com/netty/netty/issues/3726
    
    Motivation:
    
    libaio is a fast library for accessing files, implementing a JNI wrapper
    would give a boost on accessing files
    
    Modification:
    
    * Add DirectFileDescriptorController Which will handle the libaio queue among multiple files
    * Add DirectFileDescriptor representing the files
    * ErrorInfo A class to contain error information on callbacks in case of submission failures
    * Added a native module as part of transport-native-epoll
    
    Result:
    
    There is now a library that could handle files through asynchronous queues and direct memory access

netkins commented:
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

normanmaurer commented:
@netkins accept

clebertsuconic commented:
Netkins needs libaio installed for the dependencies

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.h:
```diff
@@ -0,0 +1,26 @@
+void Java_io_netty_channel_libaio_DirectFileDescriptorController_deleteContext(JNIEnv* env, jclass clazz, jobject pointer);
+void Java_io_netty_channel_libaio_DirectFileDescriptorController_submitWrite(JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferWrite, jobject callback);
+void Java_io_netty_channel_libaio_DirectFileDescriptorController_submitRead(JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferWrite, jobject callback);
+jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newAlignedBuffer(JNIEnv * env, jclass clazz, jint size, jint alignment);
```
JNICALL should be removed ?

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_unix_FileDescriptor.h:
```diff
@@ -17,3 +17,6 @@
 int Java_io_netty_channel_unix_FileDescriptor_close(JNIEnv* env, jclass clazz, jint fd);
 int Java_io_netty_channel_unix_FileDescriptor_open(JNIEnv* env, jclass clazz, jstring path);
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptor_freeContext
+  (JNIEnv *, jclass, jobject);
```
merge the two lines

clebertsuconic replied:
mistake.. I was not supposed to touch this file.. removed this

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:
```diff
@@ -43,11 +43,15 @@
+       if (!name.startsWith("linux")) {
+          throw new IllegalStateException("Only supported on Linux");
+       }
+       NativeLibraryLoader.load("netty-transport-native-epoll", PlatformDependent.getClassLoader(Native.class));
```
what is the benefit to move this to an extra method,

clebertsuconic replied:
I need to be able to call loadLibrary from DirectFileDescriptorController as well.

normanmaurer replied:
I see... Thanks!

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,132 @@
+import static io.netty.util.internal.ObjectUtil.*;
+
+/**
+ * This is an extension to {@link io.netty.channel.unix.FileDescriptor} where the file is open with O_DIRECT *
```
no need to use a FDCN here. Just use `{@link FileDescriptor}`

trustin commented:
@clebertsuconic Installed libaio-devel and libaio on slave-centos6-1. @normanmaurer, could you install it on centos6-2?

normanmaurer commented:
@trustin @clebertsuconic same here.

normanmaurer commented:
@netkins build

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,132 @@
+        // TODO: do I need to use the alignment of the file itself or not?
+        return DirectFileDescriptorController.newAlignedBuffer(size, 512);
+    }
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,132 @@
+     * It will allocate a buffer to be used on libaio operations.
+     * Buffers here are allocated with posix_memalign
+     * You need to explicitly free the buffer created from here using the
+     *    * {@link io.netty.channel.libaio.DirectFileDescriptorController#freeBuffer(java.nio.ByteBuffer)}
```
no need to use FQCN here

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+
+/**
+ *
+ * This class is used as an aggregator for the DirectFileDescriptor
```
`{@link DirectFileDescriptor}`

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+ *
+ * A Single controller can server multiple files. There's no need to create one controller per file * *
+ *
+ * Interesting reading for this: https://ext4.wiki.kernel.org/index.php/Clarifying_Direct_IO's_Semantics
```
use `<a href=....>`

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+ *
+ * You need to use the poll methods to read the result of write and read submissions
+ *
+ * You also need to use the special buffer created by DirectfileDescriptor as you need special alignments
```
`{@link DirectFileDescriptor}`

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    protected void finalize() throws Throwable {
```
`@Override` ?

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+     * Thread polling for any reason
+     *
+     * @param callbacks area to receive the callbacks passed on submission. In case of a failure you will see an
+     *                  @{link ErrorInfo} as an element. The size of this callback has to be >= max
```
`{@link ...}`

normanmaurer replied:
replace >= with valid html

clebertsuconic replied:
You lost me... what you meant by link here?

normanmaurer replied:
Replace `@{link ErrorInfo}` with *{@link ErrorInfo}`

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+     * @return Number of callbacks returned
+     *
+     * @see io.netty.channel.libaio.DirectFileDescriptor#write(long, int, java.nio.ByteBuffer, Object)
+     * @see io.netty.channel.libaio.DirectFileDescriptor#read(long, int, java.nio.ByteBuffer, Object)
```
don't use FQCN here

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+                                   Object callback) throws IOException;
+
+    /**
+     * Documented at {@link io.netty.channel.libaio.DirectFileDescriptor#read(long, int, java.nio.ByteBuffer, Object)}*
```
Not use FQCN here

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,169 @@
+    public static native void freeBuffer(ByteBuffer buffer);
+
+    /**
+     * Documented at {@link io.netty.channel.libaio.DirectFileDescriptor#write(long, int, java.nio.ByteBuffer, Object)}*
```
Not use FQCN here

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/ErrorInfo.java:
```diff
@@ -0,0 +1,57 @@
+ * In case of a failure the callback will be replaced by an instance of this class.
+ * This will be created by the native layer on the rare event of a failure. *
+ */
+public class ErrorInfo {
```
final

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/ErrorInfo.java:
```diff
@@ -0,0 +1,57 @@
+        return error;
+    }
+
+    public String getMessage() {
```
Remove get* prefixes... we not use get mostly.

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/package-info.java:
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2014 The Netty Project
```
2015

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/package-info.java:
```diff
@@ -0,0 +1,27 @@
+ * Three main classes on this package:
+ * {@link io.netty.channel.libaio.DirectFileDescriptorController} which will handle the native methods and the queue
+ * {@link io.netty.channel.libaio.DirectFileDescriptor} which represents the file itself
+ * {@link io.netty.channel.libaio.ErrorInfo} to be placed on the callback results in case of errors.
```
Not use FQCN here

normanmaurer commented on transport-native-epoll/src/main/native-package/m4/custom.m4:
```diff
@@ -0,0 +1,38 @@
+dnl ---------------------------------------------------------------------------
+dnl  Copyright 2013 The Netty Project
```
2015

normanmaurer commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,364 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This test is using a different package from {@link io.netty.channel.libaio.DirectFileDescriptor}
```
Not use FQCN here

clebertsuconic commented:
@trustin  @normanmaurer  you need libaio-devel for the libaio.h. You probably only installed the runtime.

clebertsuconic commented:
But let me rebase it first before you ask another build to Netkins

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    /**
+     */
```
remove this cruft

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+
+     * Notice: this won't hold a global reference on buffer, callback should hold a reference towards bufferWrite.
+     *         And don't free the buffer until the callback was called as this could crash the VM *
+     * *
```
remove *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+     *   {@link DirectFileDescriptorController#poll(java.nio.ByteBuffer, Object[], int, int)}* *
+
+     * Notice: this won't hold a global reference on buffer, callback should hold a reference towards bufferWrite.
+     *         And don't free the buffer until the callback was called as this could crash the VM *
```
Remove *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+    /**
+     *
+     * It will submit a write to the queue. The callback sent here will be received on the
+     *   {@link DirectFileDescriptorController#poll(java.nio.ByteBuffer, Object[], int, int)}* *
```
remove * *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+
+     * Notice: this won't hold a global reference on buffer, callback should hold a reference towards bufferWrite.
+     *         And don't free the buffer until the callback was called as this could crash the VM *
+     * *
```
remove extra *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+     *   {@link DirectFileDescriptorController#poll(java.nio.ByteBuffer, Object[], int, int)}* *
+
+     * Notice: this won't hold a global reference on buffer, callback should hold a reference towards bufferWrite.
+     *         And don't free the buffer until the callback was called as this could crash the VM *
```
remove *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+    /**
+     *
+     * It will submit a read to the queue. The callback sent here will be received on the
+     *   {@link DirectFileDescriptorController#poll(java.nio.ByteBuffer, Object[], int, int)}* *
```
remove * *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+     * It will allocate a buffer to be used on libaio operations.
+     * Buffers here are allocated with posix_memalign
+     * You need to explicitly free the buffer created from here using the
+     *    * {@link DirectFileDescriptorController#freeBuffer(java.nio.ByteBuffer)}
```
s/java.nio.ByteBuffer/ByteBuffer/

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+        DirectFileDescriptorController.submitRead(this.fd, libaioContext, position, size, buffer, callback);
+    }
+
+
```
remove extra empty line

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+        this.context = newContext(queueSize);
+    }
+
+    public void close() {
```
javadocs

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    /**
+     * It will open a new file using O_DIRECT *
```
remove *

normanmaurer replied:
missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    /**
+     * It will open a new file using O_DIRECT *
```
remove *

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+
+    /**
+     * It will poll the libaio queue for results. It should block until min is reached
+     * Results are placed on the callback*
```
remove *

normanmaurer replied:
missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+     * Results are placed on the callback*
+     *
+     * This shouldn't be called concurrently. You should provide your own synchronization if you need more than one
+     * Thread polling for any reason
```
missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+    ByteBuffer context;
+
+    /**
+     * The queue size here will use resources defined on etc. MAX_AIO *
```
remove * and missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+
+/**
+ *
+ * This class is used as an aggregator for the {@link DirectFileDescriptor}
```
missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+ *
+ * It holds native data, and it will share a libaio queue that can be used by multiple files.
+ *
+ * You need to use the poll methods to read the result of write and read submissions
```
missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+ * You need to use the poll methods to read the result of write and read submissions
+ *
+ * You also need to use the special buffer created by {@link DirectFileDescriptor} as you need special alignments
+ * when dealing with O_DIRECT files
```
missing '.' on EOL

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,172 @@
+ * You also need to use the special buffer created by {@link DirectFileDescriptor} as you need special alignments
+ * when dealing with O_DIRECT files
+ *
+ * A Single controller can server multiple files. There's no need to create one controller per file * *
```
remove * * and missing '.' on EOL

clebertsuconic commented:
What's the point of the '.' on EOL? you want . at every end of sentence?

normanmaurer commented:
Yes that is what we do



> Am 28.05.2015 um 19:39 schrieb clebertsuconic <notifications@github.com>:
> 
> What's the point of the '.' on EOL? you want . at every end of sentence?
> 
> —
> Reply to this email directly or view it on GitHub.
>

clebertsuconic commented:
@normanmaurer I believe I fixed those

clebertsuconic commented:
I rebased against the last master.. but kept the individual commits showing the progress of the reviews here

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    int used;
+};
+
+static inline struct io_control * getIOControl(JNIEnv* env, jobject pointer ) {
```
s/pointer )/pointer)/

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+static inline struct iocb * getIOCB(struct io_control * control) {
+    pthread_mutex_lock(&(control->iocbLock));
+
+    struct iocb * iocb = 0;
```
this can go over the `pthread_mutex_lock(...)`.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    struct iocb ** iocb = (struct iocb **)malloc(sizeof(struct iocb *) * queueSize);
+
+    for (i = 0; i < queueSize; i++)
+    {
```
Can you reformat the c files to always put the curly brace on the same line ? 

So like `for (...) {` . This also applies for if statements and others. Please do this in all files.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    {
+       return;
+    }
+
```
remove extra empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+    theControl->errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (theControl->errorInfoClass == NULL)
+    {
```
you need to call free(...) here to not introduce memory leaks:

```clang
free(events);
free(theControl);
```

normanmaurer replied:
and also free what was added to `iocb` before.

Scottmitch replied:
and free `libaioContext` + `iocb`?

clebertsuconic replied:
You probably missed the deleteContext method, that is called when the Controller is unloaded.

I added a small comment saying where it's all freed

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    theControl->errorInfoConstr = (*env)->GetMethodID(env, theControl->errorInfoClass, "<init>", "(Ljava/lang/Object;ILjava/lang/String;)V");
+
+    if (theControl->errorInfoConstr == NULL)
+    {
```
you need to call free(...) here to not introduce memory leaks:

```clang
free(events);
free(theControl);
```

normanmaurer replied:
also delete the created global reference

normanmaurer replied:
and also free what was added to iocb before.

clebertsuconic replied:
this is already being freed at Java_io_netty_channel_libaio_DirectFileDescriptorController_deleteContext

the iocbs are resued. you will have a pointer** containing multiple iocbs and several instances there that are also being freed.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+    struct io_control * theControl = (struct io_control *) malloc(sizeof(struct io_control));
+
+    theControl->errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
```
FindClass should only be called once during init and reused as it is expensive and slow.

clebertsuconic replied:
I didn't have a loadNative method here since you are already using it at the Native.c

I can introduce a hook here if you like, but I don't expect having more than one instance per VM (or per application if you are using muliple applications on a single VM)

clebertsuconic replied:
I'm doing this once per VM load, but I think what we had there was fine since this factory / controller was supposed to be called once per VM only.

I added a hook, what actually makes me think this library should be a separate module.

or maybe you should you should rename transport-native-epoll to just native-libs or something more generic... say you could introduce even more native stuff as you go.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_deleteContext(JNIEnv* env, jclass clazz, jobject contextPointer) {
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+}
+
+JNIEXPORT int JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_open(JNIEnv* env, jclass clazz, jstring path) {
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+        }
+        else {
+            throwIOException(env, exceptionMessage("Error while submitting IO:", -result));
+            return;
```
no need for the above two return statements

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitWrite
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferWrite, jobject callback) {
+
```
remove empty line

clebertsuconic replied:
I'm doing these as you're asking.. but I find it ugly to not have a space there.. it gets more crowded to the method declaration ;) 
I'm easy though.. just saying I would prefer a space after the method declaration.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitRead
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferRead, jobject callback) {
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    io_prep_pread(iocb, fileHandle, getBuffer(env, bufferRead), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+
```
remove extra empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+JNIEXPORT jint JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_poll
+  (JNIEnv * env, jobject obj, jobject contextPointer, jobjectArray callbacks, jint min, jint max) {
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+(JNIEnv * env, jclass clazz, jint size, jint alignment) {
+    if (size % alignment) {
+        throwRuntimeException(env, "Buffer size needs to be aligned to passed argument");
+        return 0;
```
return NULL ?

clebertsuconic replied:
It doesn't really matter, these are after a throwRuntimeException call, the java method will signal an exception and these will just be ignored.

I will set it to NULL just for consistency.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+    if (result) {
+        throwRuntimeException(env, exceptionMessage("Can't allocate posix buffer:", result));
+        return 0;
```
return NULL ?

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+        return 0;
+    }
+
+    jobject jbuffer = (*env)->NewDirectByteBuffer(env, buffer, size);
```
just return it directly ?

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+  	void *  buffer = (*env)->GetDirectBufferAddress(env, jbuffer);
+  	free(buffer);
+}
```
and CLRF for end of file

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_freeBuffer
+  (JNIEnv * env, jclass clazz, jobject jbuffer) {
+
+  	void *  buffer = (*env)->GetDirectBufferAddress(env, jbuffer);
```
remove extra ' ' after '*'

Scottmitch replied:
Do you need another temporary variable?  Unless you need to check for `NULL` (which isn't being done now) just inline:

```c
free((*env)->GetDirectBufferAddress(env, jbuffer));
```

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_freeBuffer
+  (JNIEnv * env, jclass clazz, jobject jbuffer) {
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.h:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2014 The Netty Project
```
2015

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_unix_FileDescriptor.c:
```diff
@@ -38,3 +38,4 @@ JNIEXPORT int JNICALL Java_io_netty_channel_unix_FileDescriptor_open(JNIEnv* env
     }
     return res;
 }
+
```
revert change ?

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,133 @@
+        checkNotNull(path, "path");
+
+        if (ioContext == null) {
+            throw new IOException("Could not initialize libaio context");
```
checkNotNull(ioContext, "ioContext"); ?

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,133 @@
+     * @return the buffer allocated.
+     */
+    public ByteBuffer newBuffer(int size) {
+        // TODO: do I need to use the alignment of the file itself or not?
```
Can you clarify here ? I'm not sure I understand the TODO :)

clebertsuconic replied:
You need to be aligned by 512 on libaio.

However the best really seems to be aligned to the file system alignment.

There is an API (fstat) to get the file's alignment. 
this TODO is to check if that's needed or not.

I think it's fine with 512. but need to double check as it's not clear from the little doc I could find.

normanmaurer commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,380 @@
+            Assert.assertEquals(0, count.get());
+        }
+    }
+
```
remove empty line

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+};
+
+static inline struct io_control * getIOControl(JNIEnv* env, jobject pointer ) {
+    struct io_control * ioControl = (struct io_control *) (*env)->GetDirectBufferAddress(env, pointer);
```
You can also just return it directly.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    theControl->used = 0;
+    pthread_mutex_init(&(theControl->iocbLock), 0);
+
+    jobject bufferContext = (*env)->NewDirectByteBuffer(env, theControl, sizeof(struct io_control));
```
return directly ?

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    free(theControl->iocb);
+
+    free(theControl->events);
+
```
remove empty lines ?

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * Open a new {@link FileDescriptor} for the given path.
+     */
+    static DirectFileDescriptor from(String path, ByteBuffer ioContext) throws IOException {
```
is this package private by design ?

clebertsuconic replied:
it is meant to be called from the controller only. You see any issues?

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+        // Error, so need to release whatever was done before
+        free(libaioContext);
+
+        throwRuntimeException(env, exceptionMessage("Cannot initialize queue:", -res));
```
I briefly searched online but couldn't find any details on `ThrowNew` from JNI. It looks like we are assuming this method will do a `free` on the `message` we pass it, do we have any documentation or proof that details this behavior?

clebertsuconic replied:
I'm just using whatever Norman was using already.

Scottmitch replied:
@normanmaurer - Can you comment?

clebertsuconic replied:
I don't think throwNew can possibly release anything.
I also think exceptionmessage is creating a small leak there.

I can add a fix for that with this PR.

But again: do you need to be that detailed on the exceptional cases? these exceptions will be thrown when things are already unbehaved and most likely the JVM is going down anyways.

normanmaurer replied:
@clebertsuconic yes we should fix every leak. And yes please fix it as a separate PR  so we can fix it for epoll as well in the meantime.

clebertsuconic replied:
I will add a separate commit as part of the same PR.

My PR is now dependent on this function, so I'm fixing the epoll as part of this as well

clebertsuconic replied:
@normanmaurer actually I sent another PR: https://github.com/netty/netty/pull/3844

clebertsuconic replied:
I will rebase this one on top of that (as it will be dependent on these exceptions).

Since you guys use cherry-pick it will be all good

Scottmitch replied:
Cool...that is what I thought...pre-existing leak.  just happened to notice it here...thanks for brining my attention to it :)

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * This represents a structure allocated on the native
+     * this is a io_context_t
+     * * *  */
```
nit: remove extra `*`?

clebertsuconic replied:
this was already fixed I think.

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+            throwIOException(env, "Not enough space on libaio queue");
+            return;
+        }
+        else {
```
remove `else`...you are returning in the if block above.

clebertsuconic replied:
already fixed

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    #endif
+
+    if (control->used < control->queueSize)
+    {
```
Use consistent formatting for braces.  Preferably on same line.

```java
if () {
} else {
}
```

clebertsuconic replied:
already fixed

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    pthread_mutex_unlock(&(control->iocbLock));
+}
+
+static inline void * getBuffer(JNIEnv* env, jobject pointer ) {
```
remove extra space between `pointer )`

clebertsuconic replied:
already fixed

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+    for (i = 0; i < queueSize; i++)
+    {
+       iocb[i] = (struct iocb *)malloc(sizeof(struct iocb));
```
Not sure if this will over complicate things...but is it possible to lazy allocate/free memory?  So just set to `NULL` here and allocate when needed?

clebertsuconic replied:
I don't think that's needed... these will certainly be used at a fixed buffer.
I would just keep it simple.

Most libaio usage will always alocate and free iocb all the time. I'm already buffering it what is a big improvement already.

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    theControl->errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (theControl->errorInfoClass == NULL)
+    {
+       return;
```
`return NULL`?

clebertsuconic replied:
this method was moved to static onLoad

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+    if (theControl->errorInfoConstr == NULL)
+    {
+       return;
```
`return NULL`?

clebertsuconic replied:
moot, moved already

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    theControl->iocbPut = 0;
+    theControl->iocbGet = 0;
+    theControl->used = 0;
+    pthread_mutex_init(&(theControl->iocbLock), 0);
```
check return value and for errors (then clean up memory if necessary)?

clebertsuconic replied:
I have moved the load of the class... Besides if things go wrong here the VM will probably need to be crashed.

Scottmitch replied:
What about `EPERM` and `EAGAIN`? from the http://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_mutex_init.html?

clebertsuconic replied:
I see,

I will move the pthread_mutex_init at the begining and delete the structure if failed. (as well as throw an error).

clebertsuconic replied:
fixed as well

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    int result = io_submit(ioContext, 1, &iocb);
+
+    if (result < 0) {
+        if (result == -EAGAIN) {
```
I think you need to call `DeleteGlobalRef` for `iocb->data`?

clebertsuconic replied:
It's being called on the callback. The Object should live between the submission and the callback,
otherwise you could have a VM crash if the user didn't hold the reference himself.

Scottmitch replied:
So the callback will be called even if the `io_submit` call fails?

clebertsuconic replied:
ah... good catch.. didn't think about that case.
Although there is no reason really for the io_submit at this point since I'm checking if there's enough space from the check on the iocb. 

Things like this could happen of course If things are really bad on the VM, but at that point this would be of little help.

I will coupe with failures and make the revert just in case. but I have no idea now how to force a failure here for a test check.

clebertsuconic replied:
fixed

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    int result = io_getevents(theControl->ioContext, min, max, theControl->events, 0);
+
+    for (i = 0; i < result; i++) {
+        struct iocb * iocbp = (struct iocb *)theControl->events[i].obj;
```
`theControl->events[i]` is looked up multiple times.  Consider storing a pointer to this to avoid repeating array memory lookup arithmetic.

clebertsuconic replied:
good point. I thought the compiler would optimize this but I guess I'm too used to java now.

normanmaurer replied:
haha yeah... we take all the optimisation for granted ;)

clebertsuconic replied:
@normanmaurer  I'm not sure what do here really.

Scottmitch replied:
In this case there are only 2 references so not a huge deal and the compiler may optimize this depending on your optimization level...but why leave it up to the compiler if it is easy enough to do yourself.

```c
    for (i = 0; i < result; i++) {
        strcut MYTYPE* event = &theControl->events[i]
        struct iocb * iocbp = (struct iocb *)event->obj;

        int eventResult = event->res;
        ...
    }
```

just substitute `MYTYPE` for what ever the actual type is.

clebertsuconic replied:
fixed

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+                                                   jstrError);
+
+            if (errorObject == NULL) {
+                return -1;
```
There is some cleanup code lower in the method (i.e. `DeleteGlobalRef`).  Is it OK to skip that in this situation?

clebertsuconic replied:
that's not being skipped... 

and I also have a test validating for leaks.. it would show up on that test if it wasn't

Scottmitch replied:
Can you point out where the equivalent of `(*env)->DeleteGlobalRef(env, (jobject)iocbp->data);` is being called when this condition is hit?  Or comment as to why it isn't necessary?

clebertsuconic replied:
Ah, I missed your point about this.

Do we really need to be detailed at that level here?

if this returns NULL, the JNI will have an exception thrown. Meaning: things are really bad at this point.


This would be literally: treat the exception of the exception of the exception.

An error there should be a rare event, Not having the Object available... things are really bad anyways!!!    It can't get any worse!

clebertsuconic replied:
this should be fixed now on my last rebase

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+
+JNIEXPORT jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newAlignedBuffer
+(JNIEnv * env, jclass clazz, jint size, jint alignment) {
+    if (size % alignment) {
```
Don't treat an integer arithmetic expression as a boolean expression.  Add `== 0` if this is the intention.

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,347 @@
+    // This will allocate a buffer, aligned by alignment.
+    // Buffers created here need to be manually destroyed by destroyBuffer, or this would leak on the process heap away of Java's GC managed memory
+    // NOTE: this buffer will contain non initialized data, you must fill it up properly
+    void * buffer = 0;
```
`= NULL`?

clebertsuconic replied:
it could be anything really.. I'm just delcaring this as void * buffer, non initialized.

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:
```diff
@@ -43,11 +43,15 @@
+    }
+
+    public static void loadLibrary() {
+       String name = SystemPropertyUtil.get("os.name").toLowerCase(Locale.UK).trim();
```
Is `Locale.UK` required?

normanmaurer replied:
@Scottmitch better specify the `Locale` to be safe.

Scottmitch replied:
@normanmaurer - I'm more asking about the specific choice of `Locale.UK`.  Does it have to be this specific Locale? What if this is not available, and another that may work is available (for example the default)?

clebertsuconic replied:
This is not related to accepting my PR though

trustin replied:
nit: We use 4-space indentation.

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptor.java:
```diff
@@ -0,0 +1,133 @@
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import static io.netty.util.internal.ObjectUtil.*;
```
change to explicit import?

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,166 @@
+ * A Single controller can server multiple files. There's no need to create one controller per file.
+ *
+ * Interesting reading for this: <a href="https://ext4.wiki.kernel.org/index.php/Clarifying_Direct_IO's_Semantics"/>.
+ *
```
kill empty line?

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,166 @@
+import java.nio.ByteBuffer;
+
+/**
+ *
```
kill empty line?

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/DirectFileDescriptorController.java:
```diff
@@ -0,0 +1,166 @@
+     *
+     * It is unsafe to close the controller while you have pending writes or files open as
+     * this could cause core dumps or VM crashes.
+     * *
```
kill extra `*`

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/ErrorInfo.java:
```diff
@@ -0,0 +1,67 @@
+
+    /**
+     * This will contain the original callback replaced by this failure.
+     * @return
```
kill it or fill it :)

Scottmitch commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,380 @@
+
+    @Before
+    public void setUpFactory() {
+        control = new DirectFileDescriptorController(50);
```
Anything special about `50` here?

clebertsuconic replied:
an arbitrary number I'm using on this test... I could use a constant here.

Scottmitch replied:
Not a big deal in a test...just curious if it was significant at all.

Scottmitch commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    static class CountClass {
+        static AtomicInteger count = new AtomicInteger(0);
```
`0` should be the default. consider removing.

Scottmitch commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,380 @@
+        }
+
+        public static void checkLeaks() throws InterruptedException {
+            long timeout = System.currentTimeMillis() + 5000;
```
Use `System.nanoTime()` to measure execution time and delays?

clebertsuconic replied:
I just want a timeout for 5 seconds. if the objects are not released for 5 seconds then it would throw a failure through the assertEquals. 

I think this is sufficient.

Scottmitch replied:
I think `System.nanoTime()` should be the preferred method when measuring execution time.  `System.currentTimeMillis()` is not monotonic where `System.nanoTime()` is.  See [nanoTime javadocs](http://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()) for how to use it to compute elapsed time.

clebertsuconic replied:
I am not measuring time here. I just need to wait some time really.. 5000 milliseconds will be more than enough whatever round is happening at the VM.

I will do it just to keep you happy, but it is not really needed here.

clebertsuconic replied:
fixed on the last rebase

Scottmitch replied:
@clebertsuconic - Just to clarify... I'm not concerned about precision.  I'm concerned about boarder cases such as overflow and rollback.

1. What if the system clock (or w/e clock `System. currentTimeMillis()` uses) is moved forward/backward.  In this case we could end up waiting a lot longer than we should, or not waiting at all.
2 `System. currentTimeMillis()` is based on a fixed epoch.  Eventually this will overflow.  This is a test, and chances are we won't be running during this time, but best to avoid this situation all together if possible.

clebertsuconic replied:
Fair enough.  I fixed it I think.  Let me know if you see any other concerns

clebertsuconic commented:
I have rebased all my changes and pushed -f

clebertsuconic commented:
I think I have got all covered so far... 

only thing missing now is the TODO, that I could just remove for now...

There's a separate PR on the exception allocation.. 


that's all for today.. have a nice weekend :)

clebertsuconic commented:
@normanmaurer  I will make a small modification. the submit methods to return true / false when there are no space available on the queue.

that way you could do this on certain usecases:

```java
 if (!submitWrite(.. ))  {
     // pollRoutine(....
 }
```

clebertsuconic commented:
@normanmaurer  I screwed up with rebasing things here...  let me fixup everything tomorrow

clebertsuconic commented:
@normanmaurer I have changed the implementation API so it could also cope with regular files without using O_DIRECT. I haven't measured the benefit of libaio without O_DIRECT, but I wanted to keep this implementation a direct translation of libaio in Java, so it could be used any way is needed.

I have also added the boolean on the submit methods, so you can now poll and retry if the queue is full

and I have fixed some mix up I had on Friday with rebases. I had rebased for the previous fix and I lost some of the last changes which should now be taken care of.

I mean to keep Hiram's commit separate (so it would give him authorship on the build changes), but if you could please squash this last commit on the original one?

thank you

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,378 @@
+        if (control->iocbGet >= control->queueSize) {
+           control->iocbGet = 0;
+        }
+    } else {
```
wrap the entire else block in `#ifdef DEBUG`?

clebertsuconic replied:
@Scottmitch Honestly I don't like it for Aesthetic reasons. An else inside the #ifdef... seems like the if within the else... I don't like it..  :) maybe it's my OCD kicking on here.. lol

I only needed this when developing the initial version. that case will issue a return false after my last change, I can just get rid of the entire else. you ok with that?

Scottmitch replied:
@clebertsuconic - Yes get rid of the `else` if it is not needed.

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,378 @@
+    if (res)
+    {
+        throwRuntimeExceptionErrorNo(env, "Can't initialize mutext:", res);
+        return;
```
Should we ensure we cleanup `free(libaioContext);` an other memory that is cleaned up above here?

clebertsuconic replied:
@Scottmitch  right.. I had it done on my friday night version, but I messed up something that I lost all my changes. I ended up forgetting this one... thanks for bringing attention.

Scottmitch commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,428 @@
+        public static void checkLeaks() throws InterruptedException {
+            long timeout = System.nanoTime() + TimeUnit.NANOSECONDS.convert(5, TimeUnit.SECONDS);
+
+            while (timeout > System.currentTimeMillis() && count.get() != 0) {
```
Use `nanoTime` instead of `currentTimeMillis`

Scottmitch commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,428 @@
+        }
+
+        public static void checkLeaks() throws InterruptedException {
+            long timeout = System.nanoTime() + TimeUnit.NANOSECONDS.convert(5, TimeUnit.SECONDS);
```
Also see [nanoTime javadocs](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()) for how to use nano time.

```
To compare two nanoTime values

 long t0 = System.nanoTime();
 ...
 long t1 = System.nanoTime();
one should use t1 - t0 < 0, not t1 < t0, because of the possibility of numerical overflow.
```

clebertsuconic replied:
This was a mixup I had from my version on Friday :/

clebertsuconic replied:
I will just a a simple for here with number of retries instead of calculating time

clebertsuconic commented:
@Scottmitch  https://github.com/clebertsuconic/netty/commit/ae344cb1e5a441db518ec61ddf695860dab5c83b

Scottmitch commented on transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java:
```diff
@@ -0,0 +1,425 @@
+        }
+
+        public static void checkLeaks() throws InterruptedException {
+            for (int i = 0; count.get() != 0 && i < 50; i++) {
```
@clebertsuconic - Can you comment on your observations on how high `i` typically gets before cleanup is "finished successfully"?

clebertsuconic replied:
It's just how long it takes for a full GC to finish. on my tests the best number was around 100 milliseconds. I used 50 milliseconds and it was taking 2 System.gc calls eventually.

Scottmitch replied:
@clebertsuconic - Thanks.  Just trying to get an idea of how likely this cleanup is to fail because we didn't wait long enough.

clebertsuconic replied:
If it's taking anywhere > 5 we are likely to have a leak already.
I'm using a large number on this case for cases where you are using a VM box or something.

I guess I just realized how often I'm using the number 50 on arbitrary numbers :)

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c:
```diff
@@ -370,6 +370,10 @@ static int init_in_addr(JNIEnv* env, jbyteArray address, struct in_addr* addr) {
 // util methods end
 
+// prototypes for the load hook from DirectFileDescriptorController.c
+void directFile_JNI_OnLoad(JNIEnv* env);
```
we should return `jint` here so we can return JNI_ERR if something bad happens. WDYT ?

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,388 @@
+
+    errorInfoConstr = (*env)->GetMethodID(env, errorInfoClass, "<init>", "(Ljava/lang/Object;ILjava/lang/String;)V");
+
+    if (errorInfoConstr == NULL) {
```
I think we should return JNI_ERR. See above.

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c:
```diff
@@ -0,0 +1,388 @@
+void directFile_JNI_OnLoad(JNIEnv* env) {
+    errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (errorInfoClass == NULL) {
+       return;
```
I think we should return JNI_ERR. See above.

clebertsuconic replied:
@normanmaurer 07ed66f9d5225624382d5e016461d6a8bdff4417

clebertsuconic commented:
@normanmaurer/@scottmitch  since I added a parameter to use O_DIRECT on openFile, I'm not sure the name DirectFileDescriptor describes it well now.

I was thinking if I shouldn't rename these two classes as LibaioFile and LibaioFileController. LibaioFile could still extend FileDescriptor.

I'm not sure I should do that though.. wanted a second opinion. WDYT?

clebertsuconic commented:
I am adding 1f69f1ac2b49f149518b8b46f571f879fb0e16c6 renaming these classes

We can revert it If you guys don't like it.. but I'm liking it better. (sort of an evolution of the original idea)

Scottmitch commented:
@clebertsuconic - I don't have a problem with this. @normanmaurer - WDYT?

normanmaurer commented:
@clebertsuconic @Scottmitch works for me

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_LibaioContext.c:
```diff
@@ -0,0 +1,390 @@
+}
+
+/**
+ remove an iocb from the pool of IOCBs. Returns null if full
```
isn't there a '*' a missing before 'remove' ?

clebertsuconic replied:
I didn't realize you were using javadoc style comments on Native.c

i will fix all the other ones

normanmaurer commented on transport-native-epoll/src/main/c/io_netty_channel_libaio_LibaioContext.c:
```diff
@@ -0,0 +1,390 @@
+
+        return JNI_FALSE;
+    }
+    else {
```
remove else and just return JNI_TRUE

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/libaio/package-info.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ * Three main classes on this package:
+ * {@link DirectFileDescriptorController} which will handle the native methods and the queue
+ * {@link DirectFileDescriptor} which represents the file itself
```
Fix up DirectFileDescriptor links after renaming classes

clebertsuconic replied:
@normanmaurer  29f9b0c1c24dcf02d70fdb29b1cebf4292d589e7

clebertsuconic commented:
@normanmaurer  the Netkins build failed because this box didn't have libaio-devel

What's missing for a merge now?

trustin commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:
```diff
@@ -43,11 +43,15 @@
-            throw new IllegalStateException("Only supported on Linux");
-        }
-        NativeLibraryLoader.load("netty-transport-native-epoll", PlatformDependent.getClassLoader(Native.class));
+       loadLibrary();
```
nit: We use 4-space indentation.

trustin commented:
It looks good codewise, but I wonder if netty-transport is a good place for this. Maybe we could move this into netty-tcnative for example? What do you think?

trustin commented:
BTW I asked the folks at ClinkerHQ to install libaio-dev.

trustin commented:
@netkins build

clebertsuconic commented:
@trustin I would rename transport-native-epoll into netty-native

clebertsuconic commented:
@trustin I just looked at Tomcat Native and it doesn't seem a good fit.

IMO the best idea would be to rename transport-native-epoll into netty-native (a more generic name so you would open space for other native methods you may like).

hepin1989 commented:
@clebertsuconic rename +1,but that would make a break for change the deps right?

clebertsuconic commented:
2 Options: 

- Make the rename on master(5) only.. keep the same on 4
- Add a new native method, keep the old one
- make the transport-native-epoll to mirror the new renamed method in deprecate mode

Dunno!?

trustin commented:
Interesting suggestion.  Eventually, I'd like to see netty-tcnative is replaced with it (.. if we decide to follow your suggestion.)

clebertsuconic commented:
The failures are not related to my changes. they look intermittent failures on the epoll

clebertsuconic commented:
@trustin / @normanmaurer  what should I do here?

trustin commented:
@clebertsuconic We can keep this in the epoll for now. We might want to move to a new native module like `netty-native` as you suggested in the future. I like the idea.

I wonder if the libaio package should be under io.netty.channel though. To me, it has nothing that implements `Channel`, right?  How about moving them to somewhere like io.netty.native.libaio? I'm not sure OSGi allows that, though. Thoughts?

normanmaurer commented:
@trustin "native" can not be used in a package name. But yeah I agree we should maybe better move it.

trustin commented:
@normanmaurer Haha, yeah. That's a great point. Suggestions? `io.netty.jni`?

normanmaurer commented:
@trustin yep sounds not bad :)

clebertsuconic commented:
@normanmaurer @trustin  added a commit to rename the packages (also did a few test changes on this last commit)


Please rebase all my committs into one.. Or I can after you finish your reviews.


I intend to keep Hiram's commit separate to give him authorship of the changes you prefer it differently.

clebertsuconic commented:
@normanmaurer  if you like, I could create a project jlibaio on github, and you consume it from there independently.

clebertsuconic commented:
It will make more sense for this to be an offline project.

I still don't have a home. I will certainly use it on artemis.

One option I have is to make a new organization jlibaio...

Whenever you decide to use libaio please contact me, I don't think it would make sense you creating another library for doing the same thing.


Thanks guys for the interaction here.
